### PR TITLE
Adds support for nested Router components.

### PR DESCRIPTION
**Pull request by robframpton**

### CODE REVIEW

**Comment by jbalsas** [comment]
Thanks for clarifying, @Robert-Frampton, that's what I was assuming based on the code, but better safe than sorry 😉 

So... I assume one would likely want to use this like:

```javascript
<Router path="/home" component={HomeComponent}>
  <Router path="/child" />
  <Router path="/child2" component={AnotherHomeComponent} />
</Router>

// Resulting Component Trees...

// Route: /home
<HomeComponent/>

// Route: /home/child
<HomeComponent>
    <ChildComponent />
</HomeComponent>

// Route: /home/child2
<AnotherHomeComponent/>
```

So maybe we need to inherit the component (and maybe other config) as well as the path? 

/cc @mthadley, care to weigh in?

**Comment by Robert-Frampton** [comment]
In your example, where is `ChildComponent` configured?

**Comment by jbalsas** [comment]
In my mind, it is just the result of how the component renders based on some other state. I was assuming that's how those routes work right now, but I may of course be wrong, that's why I was asking around 🤷‍♂️

**Comment by Robert-Frampton** [comment]
Oh I see, as in maybe there is a conditional in `HomeComponent` that will render the `ChildComponent` when it's at the right path/has the right state.

**Comment by jbalsas** [comment]
Yeah, taken from https://github.com/metal/metal-router/issues/14#issuecomment-264342295

```jsx
<Router component={CreateMain} path={`${PortletConstants.urls.basePath}/main/create`} >
<Router component={CreateMain} path={`${PortletConstants.urls.basePath}/main/create/:model([a-zA-Z]+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)/:action([a-zA-Z]+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)/edit/:model([a-zA-Z]+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)/edit/:model([a-zA-Z]+)/:action([a-zA-Z]+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)/edit/:model([a-zA-Z]+)/:entryId(\\d+)`} />
<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)/edit/:model([a-zA-Z]+)/:entryId(\\d+)/:action([a-zA-Z]+)`} />
<Router component={Index} path={this._defaultRoute} />
```

would turn into: 

```jsx
<Router component={CreateMain} path={`${PortletConstants.urls.basePath}/main/create`}>
  <Router path={`/:model([a-zA-Z]+)`} />
</Router>

<Router component={EditMain} path={`${PortletConstants.urls.basePath}/main/:portletMainId(\\d+)`}>
  <Router path={`/:action([a-zA-Z]+)`} />
  <Router path={`/edit/:model([a-zA-Z]+)`}>
    <Router path={`/:action([a-zA-Z]+)`} />
    <Router path={`/:entryId(\\d+)`}>
      <Router path={`/:action([a-zA-Z]+)`} />
    </Router>
  </Router>
</Router>

<Router component={Index} path={this._defaultRoute} />
```

**Comment by Robert-Frampton** [comment]
Makes sense to me. So essentially it only inherits the parent component if it doesn't have one defined. I'll go ahead add this feature.

**Comment by Robert-Frampton** [comment]
Something else we should figure out is nested routes for Soy and vanilla js implementations.

Currently this branch only enables it for JSX implementations since it's the only syntax that has "built in" child components. I think in addition to that, we can provide a new config property that is a tree of routes that will be consumed in a very similar way to the child components.

```jsx
<Router path="/home" component={HomeComponent}>
  <Router path="/child" component={ChildComponent}/>
</Router>

// Will behave the same as

Component.render(Router, {
  path: '/home',
  component: HomeComponent,
  childRoutes: [
    {
      component: ChildComponent,
      path: '/child'
    }
  ]
});
```

**Comment by mthadley** [comment]
Hey everyone, this is a really cool step in the right direction. It should help to clean up a lot of our route configurations.

So a few comments on the API:

--

At first, I would have expecte `component` to be a required prop. At least for JSX, I don't think this makes sense to me:
```jsx
<Router path="/home" component={HomeComponent}>
  <Router path="/child" />
</Router>
```
Theoretically, whether `/home` or `/home/child` is matched, it's the same component being rendered. We aren't passing any props to it either, other than I suppose the matched route. For this component to then take a different logical path of execution, it would have to then examine the route (aka, you know, routing), which to me defeats the purpose of our above `Router`s.

However, if the intention is for `/home/child` to be more like an **alias**, then I can see the use case. In other words, the developer wants `/home` and `/home/child` to render the exact same thing. **So for this reason, I think `component` being optional makes sense.**

So the `<ChildComponent />` that @Robert-Frampton mentioned, I wouldn't expect to see that.

--

This other snippet, I would expect to work like this:
```jsx
<Router path="/home" component={HomeComponent}>
  <Router path="/child" />
  <Router path="/child2" component={AnotherHomeComponent} />
</Router>

// Resulting Component Trees...

// Route: /home
<HomeComponent />

// Route: /home/child
<HomeComponent />

// Route: /home/child2
<HomeComponent>
  <AnotherHomeComponent />
</HomeComponent>
```
So basically `AnotherHomeComponent` is passed to `HomeComponent` as a child.

In general, **I would expect every matched route to have it's `component` passed as a _child_ to it's _nearest ancestor's_ `component`.**

**Comment by Robert-Frampton** [comment]
What if you don't want `AnotherHomeComponent` wrapped in anything? Does React provide a way to do that?

**Comment by Robert-Frampton** [comment]
Also in regards to not having a component defined, I think the assumption is that the data would be different between the two paths. Or maybe the `fetchUrl`.

```jsx
<Router data={dataSet1} path="/home" component={HomeComponent}>
  <Router data={dataSet2} path="/child" />
</Router>
```

**Comment by mthadley** [comment]
@Robert-Frampton I'm thinking you would omit `component` on the parent route:
```jsx
<Router path="/home">
  <Router path="/child" component={AnotherHomeComponent} />
</Router>
```

So given the matched path down the tree of `Router` components, the rendered output would essentially form a chain of components from the root down to the leaf. If any node (`Router`) doesn't have a `component`, then it doesn't add to the rendered output. It's just there as a matcher:

```jsx
<Router path="/">
  <Router path="/home" component={Home} />
  <Router path="/user" component={User}>
    <Router path="settings">
      <Router path="/permissions" component={Permissions} />
      <Router path="/email" component={Email} />
    </Router>
    <Router path="/about" component={About} />
    <Router path="/dashboard" component={Dashboard}>
      <Router path="/news" component={News} />
    </Router>
  </Router>
</Router>
```
`->`

| Path | Rendered |
| -----|-----------|
| `/home` | `Home`
| `/user` | `User`
| `/user/settings/permissions` | `User -> Permissions` |
| `/user/settings/permissions` | `User -> Email` |
| `/user/about` | `User -> About` |
| `/user/dashboard/news` | `User -> Dashboard -> News` |

This makes sense to me, what do you all think? @Robert-Frampton @jbalsas 

--

And to your other comment Rob, I agree. There would need to be something else different in that `Router`'s configuration for it to make sense.

**Comment by Robert-Frampton** [comment]
It makes sense to me. I do think it might make it more restrictive as far as a project's component architecture, and there are a few things regarding passing data that I'm unsure about.

For example.

```jsx
const data = {
	key: 'value'
};

<Router path="/user" component={User} data={data}>
	<Router path="/settings" component={Settings}>
		<Router path="/permissions" component={Permissions} />
	</Router>
</Router>
```

The parent `/user` route will receive the `data` object, but what about `/user/settings` and `/user/settings/permissions`? We could just pass all parent data down to the nested components, but that might mean passing data that is unnecessary or unintended for the children.

**Comment by mthadley** [comment]
Hmm, my first reaction is that the `User` component would be responsible for passing down any data that it received, which is the normal data flow outside of this router configuration.

But since it's only receiving `this.props.children`, it may be awkward.

**Comment by jbalsas** [comment]
Hey @mthadley, @Robert-Frampton, I think this is going in a bit too deep. I'd like to just take a little step here, but I've already shared my concerns with Rob that if we push a feature out it might hinder future development.

As this is not currently stopping anyone, I'd rather just pause it and start a discussion in an issue to figure out what exactly do we want for this. Currently, it's also hard for me to wrap my head around soy usage of this, which seems pretty much `jsx` oriented...

What do you think?

**Comment by mthadley** [comment]
@jbalsas Yea, that's fair. I should have saved it for a separate issue, instead of side tracking things. 👍 

I'm not sure about the soy usage either, but then again, it's not what I'm working with day to day. But I also know what API I'd really like to have.

**Comment by jbalsas** [comment]
Closing for now as we're archiving this repo in favour of https://github.com/metal/metal-plugins

We will pick up the conversation there and resend this once it's more clear.
